resolver/resolver.cc

```diff
@@ -295,10 +301,52 @@ class ResolveConstantsWalk {
         return !checker.seenUnresolved;
     }
 
-    static core::SymbolRef resolveConstant(core::Context ctx, const shared_ptr<Nesting> &nesting,
-                                           const ast::UnresolvedConstantLit &c, bool &resolutionFailed) {
+    // Find load-time out-of-order references.
+    // Algorithm:
+    // if resolved symbol is only defined in the current file, and
+    //    the definition is after the reference loc,
+    // then, report an error.
+    static void
+    checkReferenceOrder(core::Context ctx, core::SymbolRef resolutionResult, const ast::UnresolvedConstantLit &c,
+                        const shared_ptr<UnorderedMap<core::SymbolRef, core::LocOffsets>> firstDefinitionLocs) {
+        if (!ctx.file.data(ctx).isRBI() && resolutionResult.exists() &&
+            resolutionResult.isOnlyDefinedInFile(ctx.state, ctx.file)) {
+            core::LocOffsets defLoc;
+            const auto it = firstDefinitionLocs.get()->find(resolutionResult);
+
+            if (it != firstDefinitionLocs.get()->end()) {
+                // Use the loc from the local first defs table, if it exists.
+                // When the symbol is declared multiple times in the same file, we want to ensure that
+                // we compare against the *first* definition.
+                defLoc = it->second;
+            } else {
+                // If the symbol isn't in the local first defs table, we fall back to the
+                // canonical loc.
+                defLoc = resolutionResult.loc(ctx).offsets();
+            }
+
+            // Check for ordering between reference loc and definition loc.
+            const auto &refLoc = c.loc;
+            if (defLoc.beginPos() >= refLoc.endPos()) {
+                if (auto e = ctx.beginError(c.loc, core::errors::Resolver::OutOfOrderConstantAccess)) {
+                    e.setHeader("`{}` referenced before it is defined", resolutionResult.show(ctx));
+                    e.addErrorLine(ctx.locAt(defLoc), "Defined here");
+                }
+            }
+        }
+    }
+
+    static core::SymbolRef
+    resolveConstant(core::Context ctx, const shared_ptr<Nesting> &nesting, const ast::UnresolvedConstantLit &c,
+                    bool &resolutionFailed, bool loadTimeScope,
+                    const shared_ptr<UnorderedMap<core::SymbolRef, core::LocOffsets>> firstDefinitionLocs) {
```

Comment:
I vote for either taking this parameter by reference, or changing it to `const UnorderedMap<core::SymbolRef, core::LocOffsets> &` and changing the callers.

Slight preference for just taking this type by reference so we don't incur memory traffic in the `!loadTimeScope` case (which I guess should be uncommon anyway?).